src/storage/src/storage_state/async_storage_worker.rs

```diff
@@ -0,0 +1,166 @@
+    /// Creates a new [`AsyncStorageWorker`].
+    ///
+    /// IMPORTANT: The passed in `activatable` is activated when new responses
+    /// are added the response channel. It is important to not sleep the thread
+    /// that is reading from this via [`try_recv`](Self::try_recv) when
+    /// [`is_empty`](Self::is_empty) has returned `false`.
+    pub fn new<A: Activatable + Send + 'static>(
+        activatable: A,
+        persist_clients: Arc<Mutex<PersistClientCache>>,
+    ) -> Self {
+        let (command_tx, mut command_rx) = mpsc::unbounded_channel::<(tracing::Span, _)>();
+        let (response_tx, response_rx) = crossbeam_channel::unbounded();
+
+        mz_ore::task::spawn(|| "AsyncStorageWorker", async move {
+            while let Some(cmd) = command_rx.recv().await {
+                // Peel off all available commands. This allows us to catch up
+                // if we fall behind.
+                let mut commands = vec![cmd];
+                while let Ok(cmd) = command_rx.try_recv() {
+                    commands.push(cmd);
+                }
+
+                for (span, command) in commands {
+                    match command {
+                        AsyncStorageWorkerCommand::CalculateResumeFrontier(
+                            id,
+                            ingestion_description,
+                            _phantom_data,
+                        ) => {
+                            let mut persist_clients = persist_clients.lock().await;
+                            let mut state = ingestion_description
+                                .initialize_state(&mut persist_clients)
+                                .instrument(span.clone())
+                                .await;
+                            let resume_upper: Antichain<T> = ingestion_description
+                                .calculate_resumption_frontier(&mut state)
+                                .instrument(span)
+                                .await;
+                            let res = response_tx.send(
+                                AsyncStorageWorkerResponse::IngestDescriptionWithResumeUpper(
+                                    id,
+                                    ingestion_description,
+                                    resume_upper,
+                                ),
+                            );
+                            activatable.activate();
+
+                            match res {
+                                Ok(_) => {
+                                    // All's well!
+                                }
+                                Err(_err) => {
+                                    // Receiver must have hung up.
+                                    break;
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+            tracing::trace!("shutting down async storage worker task");
```

Comment:
is this not `halt`-worthy?

Reply:
not sure, in the past halting or panicing has turned out to be a bit over eager. I'll mull it over

Reply:
yeah, i think its possible its fine, if the send side of the channel is gracefully shutting down...